<a href="https://colab.research.google.com/github/anacarolinamartins07/Algorithms_for_rent_prediction/blob/main/Aluguel_Gradiente_descendente_Estoc%C3%A1stico.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Declaração do Problema
Objetivo: Prever o preço do aluguel (valor_aluguel) de propriedades residenciais.

Descrição do Conjunto de Dados:
O conjunto de dados consiste em informações sobre várias propriedades residenciais, incluindo sua localização, atributos físicos e custos associados. Cada linha representa um anúncio de propriedade com vários recursos como: area, quartos, mobiliaria, vagas_garagem dentre outras informações. Nessa previsão iremos construir uma rede reural usando o gradiente descendente estocástico.

O gradiente descendente estocástico é um algoritmo de otimização usado para treinar modelos de aprendizado de máquina. Ele atualiza os parâmetros do modelo iterativamente, minimizando a função de perda, calculando o gradiente da função de perda em relação aos parâmetros. A abordagem estocástica envolve a atualização dos parâmetros em lotes de dados de treinamento selecionados aleatoriamente, em vez de usar todo o conjunto de dados de uma vez, tornando-o computacionalmente eficiente e permitindo uma convergência mais rápida.

In [ ]:
#importando as bibiotecas
import numpy as np
import pandas as pd
import tensorflow as tf
tf.__version__ #ver versao do tensorflow

'2.15.0'

### Coleta e análise de Dados

In [ ]:
#importando conjunto de dados

df= pd.read_csv("/content/aluguel.csv", sep= ";")
df.head()

cidade  area  quartos  banheiros  vagas_garagem  andar  \
0     São Paulo    70        2          1              1      7   
1     São Paulo   320        4          4              0     20   
2  Porto Alegre    80        1          1              1      6   
3  Porto Alegre    51        2          1              0      2   
4     São Paulo    25        1          1              0      1   

   aceita_animal  mobiliaria  valor_condominio  valor_aluguel  valor_iptu  \
0              1           1              2065           3300         211   
1              1           0              1200           4960        1750   
2              1           0              1000           2800           0   
3              1           0               270           1112          22   
4              0           0                 0            800          25   

   valor_seguro  valor_total  
0            42         5618  
1            63         7973  
2            41         3841  
3            17         1421  
4            11          836

### Pré-Processamento

In [ ]:
x = df[["area", "quartos", "mobiliaria"]].values #para pegar as features
y= df["valor_aluguel"].values # .values: transforma de df para arrays fazer isso apenas para rn pois sao dados nao estruturados

### Modelagem

In [ ]:
from sklearn.model_selection import train_test_split #validacao hold out
xtreino, xteste, ytreino, yteste = train_test_split(x , y, test_size= 0.2, random_state= 20)

### Construção de uma rede neural

In [ ]:
### definindo minha entrada (Iniciando minha ANN (input da minha rede))

ann= tf.keras.models.Sequential()

In [ ]:
#criando minha primeira camada

ann.add(tf.keras.layers.Dense(units= 6, activation = "relu")) #definindo quantos neuroninos (units) e a funcao de ativacao (sigmoid), layers(camadas)

In [ ]:
#criando minha primeira camada

ann.add(tf.keras.layers.Dense(units= 6, activation = "relu")) #definindo quantos neuroninos (units) e a funcao de ativacao (sigmoid), layers(camadas)

In [ ]:
#definindo minha saida (output)

ann.add(tf.keras.layers.Dense(units= 1)) #definindo quantos neuroninos (units), layers(camadas)

### Treinando minha rede de neural

In [ ]:
from keras.optimizers import SGD #importando a biblioteca estocastico gradiente descendente

sgd= SGD(learning_rate= 0.02) #definindo minha sgd, learning_rate=(taxa de aprendizagem)

In [ ]:
ann.compile(loss= "mean_squared_error", optimizer= sgd) #mean_squad_error = calcula o (y hat), para encontrar o melhor peso sgd(gradiente descendente)

In [ ]:
#treinando a rede

ann.fit(xtreino, ytreino, batch_size= 20, epochs= 120) #batch_size faz o calculo por blocos, epochs quantas vezes ele vai repetir os calculos

Epoch 1/120
428/428 [==============================] - 1s 2ms/step - loss: 13905307868400868111941632.0000
Epoch 2/120
428/428 [==============================] - 1s 1ms/step - loss: 160993120.0000
Epoch 3/120
428/428 [==============================] - 1s 2ms/step - loss: 11555360.0000
Epoch 4/120
428/428 [==============================] - 1s 2ms/step - loss: 11547637.0000
Epoch 5/120
428/428 [==============================] - 1s 3ms/step - loss: 11552638.0000
Epoch 6/120
428/428 [==============================] - 1s 3ms/step - loss: 11551119.0000
Epoch 7/120
428/428 [==============================] - 1s 2ms/step - loss: 11557316.0000
Epoch 8/120
428/428 [==============================] - 1s 2ms/step - loss: 11553487.0000
Epoch 9/120
428/428 [==============================] - 1s 2ms/step - loss: 11558415.0000
Epoch 10/120
428/428 [==============================] - 1s 2ms/step - loss: 11555317.0000
Epoch 11/120
428/428 [==============================] - 1s 2ms/step - loss: 11547356.0000


### Validação

In [ ]:
y_predito = ann.predict(xteste)

67/67 [==============================] - 0s 1ms/step


In [ ]:
y_predito

array([[3717.109],
       [3717.109],
       [3717.109],
       ...,
       [3717.109],
       [3717.109],
       [3717.109]], dtype=float32)

#### Na nossa rede neural desenvolvemos duas camadas, cada uma com 6 neurônios, e com a mesma função de ativação que foi a "relu" (A função ReLU ativa uma unidade neural se o valor de entrada for positivo e a desativa se for negativo), nossas features foram as mesmas das previsões que fizemos com dois algoritmos diferentes em machine learnig : area, quartos, mobiliada. Vamos para as comparações entre os algoritmos de machine learning e a nossa rede neural:


1.   Resultado da previsão com regressão linear: 7374984.32
2.   Resultado da previsão com KNN: 5293210.41
3.   Resultado da previsão com a Rede neural utilizando gradiente descendente: 3717.109

### Neste caso se formos comparar o desempenho e a previsão dos algoritmos de Machine learning e a rede neural que criamos, a previsão desta rede neural usando o gradiente descendente estocástico foi melhor.



